### ISSUE TITLE
Multiple user nodes created for new user

### ISSUE THREAD

**Issue by georgwittberger** [KnowledgeAssets/myskills-server#7]
In rare cases there is the possibility that a new user - authenticating for the first time - could be created multiple times in the database. This leads to the state that no further authentication is possible because the Spring Data repository throws an exception complaining that multiple results were found for the given user name.
The reason for this behavior is that the creation of new users during authentication is not thread-safe, see class `UserCreatingUserAuthenticationConverter`. Two requests running in parallel can go into the branch where a new user is created.
The user creation itself cannot be made thread-safe because this scenario could even happen in a clustered environment.
A possible solution would be to set up a unique constraint on the "userName" property of "User" entities in the Neo4j database. Additionally, the automatic user creation must deal with the scenario that a constraint violation exception is thrown. It can recover from that case by simply loading the user one more time because it should be available in the database then.

**Comment by georgwittberger** [comment]
Closing. Basically works now if Neo4j auto-indexing is enabled. Otherwise the unique constraint is not created in the database and the problem may occur again. I will file another issue for a general solution for index and constraint creation at server startup.
